File: MapViewerEngine/wwwroot/js/timeline.js
import anime from './anime.es.js';

export function create(animations) {
    var tl = anime.timeline({
        easing: 'linear',
        loop: true,
        autoplay: false
    });
    
    for (let i = 0; i < animations.length; i++) {
        tl.add(animations[i], 0);
    }

    return tl;
}

export function createAnimation(target, keyframes) {
    return {
        targets: target,
        easing: 'linear',
        keyframes: keyframes,
        autoplay: false
    }
}

export function createKeyframeVec(duration, x, y, z) {
    return { duration: duration, x: x, y: y, z: z };
}

export function createKeyframeQuat(duration, x, y, z, w) {
    return { duration: duration, x: x, y: y, z: z, w: w };
}

export function play(timeline) {
    timeline.play();
}

export function pause(timeline) {
    timeline.pause();
}

File: MapViewerEngine/Modules/Timeline.cs
using System.Runtime.InteropServices.JavaScript;

namespace MapViewerEngine.Modules;

public partial class Timeline
{
    [JSImport("create", nameof(Timeline))]
    public static partial JSObject Create(JSObject[] animations);
    
    [JSImport("createAnimation", nameof(Timeline))]
    public static partial JSObject CreateAnimation(JSObject target, JSObject[] keyframes);

    [JSImport("createKeyframeVec", nameof(Timeline))]
    public static partial JSObject CreateKeyframeVec(int duration, double x, double y, double z);

    [JSImport("createKeyframeQuat", nameof(Timeline))]
    public static partial JSObject CreateKeyframeQuat(int duration, double x, double y, double z, double w);

    [JSImport("play", nameof(Timeline))]
    public static partial void Play(JSObject timeline);

    [JSImport("pause", nameof(Timeline))]
    public static partial void Pause(JSObject timeline);
}
